File: casa_guido/app/src/main/java/br/com/casa_guido/models/Paciente.kt
package br.com.casa_guido.models

import androidx.compose.runtime.Immutable
import kotlinx.serialization.Serializable
import java.util.UUID

@Serializable
@Immutable
data class Paciente(
    val id: String = UUID.randomUUID().toString(),
    val pessoa: Pessoa = Pessoa(),
    val status: Boolean = false,
    val nomeMae: String = "",
    val nomePai: String = "",
    val nomeOutro: String = "",
    val cirurgias: List<Cirurgia> = emptyList(),
    val quimios: List<Quimio> = emptyList(),
    val radios: List<Radio> = emptyList(),
    val remuneracao: String = "",
    val bpc: Int? = null,
    val valorBpc: String = "",
    val aptoReceberBpc: Int? = null,
    val tipoEscola: Int? = null,
    val escolaNome: String = "",
    val tamRoupa: String = "",
    val tamCalcado: String = "",
    val diagnostico: String = "",
    val origen_info_ong: String = "",
    val alterado: Boolean = false,
    val ubs: Ubs = Ubs(),
    val cras: Cras = Cras(),
    val observacoes: Array<String> = emptyArray(),
    val responsavel: Pessoa = Pessoa(),
    val conjugeResponsavel: Pessoa = Pessoa(),
    val outroResponsavel: Pessoa = Pessoa(),
    val doenca: List<DoencaDTO> = emptyList(),
    val tratamento: List<Tratamento> = emptyList(),
    val profissionalResponsavel: ProfissionalResponsavel = ProfissionalResponsavel(),
    val composicaoFamiliar: List<ComposicaoFamiliar> = emptyList(),
    val historicoSaude: HistoricoSaude = HistoricoSaude(),
    val situacaoHabitacional: SituacaoHabitacional = SituacaoHabitacional(),
    val arquivos: List<Arquivo> = emptyList()
)

@Serializable
data class PacienteRequestDTO(
    val pessoa: PessoaDTO,
    val status: Boolean,
    val nomeMae: String,
    val nomePai: String,
    val nomeOutro: String,
    val cirurgias: List<CirurgiaDTO>,
    val quimios: List<QuimioDTO>,
    val radios: List<RadioDTO>,
    val remuneracao: String,
    val bpc: Int? = null,
    val valorBpc: String,
    val aptoReceberBpc: Int? = null,
    val tipoEscola: Int? = null,
    val escolaNome: String,
    val tamRoupa: String,
    val tamCalcado: String,
    val diagnostico: String,
    val origen_info_ong: String,
    val ubs: Ubs,
    val cras: Cras,
    val observacoes: Array<String>,
    val responsavel: PessoaDTO,
    val conjugeResponsavel: PessoaDTO,
    val outroResponsavel: PessoaDTO,
    val doenca: List<DoencaDTO>,
    val tratamento: List<TratamentoDTO>,
    val profissionalResponsavel: ProfissionalResponsavelDTO,
    val composicaoFamiliar: List<ComposicaoFamiliarDTO>,
    val historicoSaude: HistoricoSaudeDTO,
    val situacaoHabitacional: SituacaoHabitacionalDTO
)

fun Paciente.toRequestDTO(): PacienteRequestDTO {
    return PacienteRequestDTO(
        pessoa = pessoa.toRequestDTO(),
        status = status,
        nomeMae = nomeMae,
        nomePai = nomePai,
        nomeOutro = nomeOutro,
        cirurgias = cirurgias.map { it.toRequestDTO() },
        quimios = quimios.map { it.toRequestDTO() },
        radios = radios.map { it.toRequestDTO() },
        remuneracao = remuneracao,
        bpc = bpc,
        valorBpc = valorBpc,
        aptoReceberBpc = aptoReceberBpc,
        tipoEscola = tipoEscola,
        escolaNome = escolaNome,
        tamRoupa = tamRoupa,
        tamCalcado = tamCalcado,
        diagnostico = diagnostico,
        origen_info_ong = origen_info_ong,
        ubs = ubs,
        cras = cras,
        observacoes = observacoes,
        responsavel = responsavel.toRequestDTO(),
        conjugeResponsavel = conjugeResponsavel.toRequestDTO(),
        outroResponsavel = outroResponsavel.toRequestDTO(),
        doenca = doenca,
        tratamento = tratamento.map { it.toRequestDTO() },
        profissionalResponsavel = profissionalResponsavel.toRequestDTO(),
        composicaoFamiliar = composicaoFamiliar.map { it.toRequestDTO() },
        historicoSaude = historicoSaude.toRequestDTO(),
        situacaoHabitacional = situacaoHabitacional.toRequestDTO()
    )
}

File: backend/models/Paciente.js
const mongoose = require('mongoose');

// Subschema de Endereço
const EnderecoSchema = new mongoose.Schema({
  cep: String,
  logradouro: String,
  numero: String,
  complemento: String,
  unidade: String,
  bairro: String,
  localidade: String,
  uf: String,
  estado: String,
  regiao: String,
  referencia: String,
  pais: { type: String, default: 'Brasil' }
}, { _id: false });

// Subschema de Pessoa (obrigatória - para o paciente)
const PessoaSchema = new mongoose.Schema({
  nome: { type: String, required: true },
  telefone: String,
  dataNascimento: String,
  cpf: { type: String, unique: true, sparse: true },
  rg: { type: String, unique: true, sparse: true },
  naturalidade: String,
  genero: String,
  escolaridade: Number,
  serie: Number,
  estadoCivil: Number,
  situacaoProfissional: Number,
  salario: String,
  cartaoSus: String,
  respPrincipal: Number,
  profissao: String,
  endereco: EnderecoSchema
}, { _id: false });

// Subschema de Pessoa Opcional (para responsáveis)
const PessoaOpcionalSchema = new mongoose.Schema({
  nome: String, // Não é required
  telefone: String,
  dataNascimento: String,
  cpf: String, // Removido unique e sparse para responsáveis
  rg: String, // Removido unique e sparse para responsáveis
  naturalidade: String,
  genero: String,
  escolaridade: Number,
  serie: Number,
  estadoCivil: Number,
  situacaoProfissional: Number,
  salario: String,
  cartaoSus: String,
  respPrincipal: Number,
  profissao: String,
  endereco: EnderecoSchema
}, { _id: false });

// Subschema de Cirurgia
const CirurgiaSchema = new mongoose.Schema({
  nome: { type: String, required: true },
  data: { type: String, required: true },
  cid: { type: String },
  observacoes: { type: String }
}, { _id: false });

// Subschema de Composição Familiar
const ComposicaoFamiliarSchema = new mongoose.Schema({
  nome: String,
  parentesco: String,
  dataNascimento: String,
  estudaOpcional: Number,
  serie: String,
  trabalhaOpcional: Number,
  renda: String
}, { _id: false });

// Subschema de Histórico de Saúde
const HistoricoSaudeSchema = new mongoose.Schema({
  doencasFamilia: [Number],
  medicamentosUsoContinuo: String,
  localProcuraAtendimento: [Number],
  recebeBeneficio: Number,
  beneficioDescricao: String
}, { _id: false });

// Subschema de Situação Habitacional
const SituacaoHabitacionalSchema = new mongoose.Schema({
  comoAdquiriuCasa: Number,
  area: Number,
  numeroComodos: Number,
  material: Number,
  bens: [Number],
  meioDeTransporte: Number,
  meioDeComunicao: Number,
  possuiBanheiros: Number
}, { _id: false });

// Subschema de Quimioterapia
const QuimioterapiaSchema = new mongoose.Schema({
  dataInicio: { type: String, required: true },
  dataUltimaSessao: { type: String },
  observacoes: { type: String }
}, { _id: false });

// Subschema de Radioterapia
const RadioTerapiaSchema = new mongoose.Schema({
  dataInicio: { type: String, required: true },
  dataUltimaSessao: { type: String },
  observacoes: { type: String }
}, { _id: false });

// Subschema de UBS
const UbsSchema = new mongoose.Schema({
  municipio: String,
  bairro: String
}, { _id: false });

// Subschema de CRAS
const CrasSchema = new mongoose.Schema({
  municipio: String,
  bairro: String
}, { _id: false });

// Subschema de Profissional Responsável
const ProfissionalResponsavelSchema = new mongoose.Schema({
  nome: String,
  crm: String
}, { _id: false });

// Schema principal do Paciente
const PacienteSchema = new mongoose.Schema({
  // Dados da pessoa principal (paciente) - OBRIGATÓRIO
  pessoa: { type: PessoaSchema, required: true },
  
  // Dados básicos do paciente
  status: Boolean,
  nomeMae: String,
  nomePai: String,
  nomeOutro: String,
  recebeRemuneracao: Number,
  remuneracao: String,
  recebeBpc: Number,
  valorBpc: String,
  diagnostico: String,
  profissionalResponsavel: ProfissionalResponsavelSchema,
  escolaNome: String,
  anoEscolar: String,
  tamRoupa: String,
  tamCalcado: String,
  origenInfoOng: String,
  idBackend: String,
  alterado: { type: Boolean, default: false },
  observacoes: [String],
  
  // Responsáveis (OPCIONAIS) - usando PessoaOpcionalSchema
  responsavel: PessoaOpcionalSchema,
  conjugeResponsavel: PessoaOpcionalSchema,
  outroResponsavel: PessoaOpcionalSchema,
  
  // UBS e CRAS (agora como subschemas)
  ubs: UbsSchema,
  cras: CrasSchema,
  
  // Histórico médico (agora como subschemas)
  cirurgias: [CirurgiaSchema],
  quimioterapias: [QuimioterapiaSchema],
  radioterapias: [RadioTerapiaSchema],
  historicoSaude: HistoricoSaudeSchema,
  
  // Dados socioeconômicos (agora como subschemas)
  composicaoFamiliar: [ComposicaoFamiliarSchema],
  situacaoHabitacional: SituacaoHabitacionalSchema
  
}, { timestamps: true });

module.exports = mongoose.model('Paciente', PacienteSchema); 